### PR TITLE
Multiple same-type database functionality

### LINKED ISSUE #63: Supporting multiple same-system databases
Currently the way we parse the `config` file, it doesn't support multiple instances of the same database system. 
The user can't provide multiple `postgres` databases for example to be replicated, because the sections will have overlap
and the previous database information will be overwritten.

Maybe one possible solution is to save each database information on an array. For example"
```c
typedef struct config_t {
	postgres_t *postgres_config;
	smtp_t *smtp_config;
	general_t *general_config;
} config_t;
``` 
instead of this, we implement it like this:
```c
typedef struct config_t {
	postgres_array_t **postgres_config;
	smtp_t *smtp_config;
	general_t *general_config;
} config_t;
```

The size of the array could be malloc'ed at runtime after we parse the `.ini` file. This of course will require some 
refactoring from our side (on `db.c` for example), therefore I believe should be given high priority to avoid refactoring
larger parts of the codebase in the future.

### PR DESCRIPTION
**Pull request by panosfol**
Closes #63

### CODE REVIEW

**Comment by panosfol** [comment]
Convert to draft since there are a few errors indicated by valgind/helgrind.
Waiting for input from @charmitro .

**Review by charmitro — APPROVED** [review]
I generally approve this but we need to find a way to make the rust email library thread-safe also.

File: configs/test.ini
[the remark this answers is not in the document]
Comment (charmitro) [reply configs/test.ini:1]: We still don't have a seperated config for this. I would also like to see a directory called `configs/samples/` where multiple configs for testing are to be stored there.

File: src/db.c [hunk 1]
```diff
@@ -37,26 +39,24 @@ void *db_operation_thread(void *arg)
 int execute_db_operations(void)
 {
 	int ret = 0;
-	pthread_t threads[db_ops_counter];
+	pthread_t threads[MAX_AVAILABLE_DBS];
 	init_db_func init_function;
 
+	// This is needed to access the localtime in a thread-safe manner
+	tzset();
```
[the remark this answers is not in the document]
Comment (charmitro) [reply src/db.c:46]: Ping.

File: configs/test.ini
[the remark this answers is not in the document]
Comment (panosfol) [reply configs/test.ini:1]: What would be the difference of the configs for testing from the configs that are already at `configs/`?

[the remark this answers is not in the document]
Comment (charmitro) [reply configs/test.ini:1]: Forget the *samples* keyword, use `tests` instead.

* `configs/example.ini` - an example config with or without predefined values
* `configs/tests/<filename>` - the rest we use for testing.

[the remark this answers is not in the document]
Comment (panosfol) [reply configs/test.ini:1]: Also, is there reason to believe our rust module is not safe in its current iteration?

**Comment by charmitro** [comment]
I'm getting weird helgrind output when running with email enabled, can 
you enable it properly (actually send an email) and see hellgrind's output?

**Comment by panosfol** [comment]
> I'm getting weird helgrind output when running with email enabled, can you enable it properly (actually send an email) and see hellgrind's output?

Yes there are actually a lot of errors, all of them from the rust module. Also the actually sending of the email does not always happen for all the databases that were replicated, indicating that there is a conflict on the threads, making the outcome unpredictable.

**Comment by charmitro** [comment]
> > I'm getting weird helgrind output when running with email enabled, can you enable it properly (actually send an email) and see hellgrind's output?
> 
> Yes there are actually a lot of errors, all of them from the rust module. Also the actually sending of the email does not always happen for all the databases that were replicated, indicating that there is a conflict on the threads, making the outcome unpredictable.

How should we handle this?
1. Should we postpone the merging of this PR until fixed? 
2. Disable email for more than one databases defined in the config?

**Comment by panosfol** [comment]
> > > I'm getting weird helgrind output when running with email enabled, can you enable it properly (actually send an email) and see hellgrind's output?
> > 
> > 
> > Yes there are actually a lot of errors, all of them from the rust module. Also the actually sending of the email does not always happen for all the databases that were replicated, indicating that there is a conflict on the threads, making the outcome unpredictable.
> 
> How should we handle this?
> 
> 1. Should we postpone the merging of this PR until fixed?
> 2. Disable email for more than one databases defined in the config?

Lets postpone merging this, and look into multithreading in the rust library.
Also we probably should create a seperate issue for this, since its about the rust library.
All branch off of this branch and start looking into it.

File: src/db.c [hunk 2]
```diff
@@ -2,6 +2,7 @@
 #include "db/postgres.h"
 #include "log.h"
 
+#include <bits/pthreadtypes.h>
```
Comment (charmitro) [inline src/db.c:5]: What's this for?

**Pull request closed without merging**